### ISSUE TITLE
Reshape before Softmax leads runtime error:Unnecessary dynamic-sized tensors

### ISSUE THREAD

**Issue by Robird** [tensorflow/tensorflow#46725]
### 1. System information

- OS: Win10 & Ubuntu 1804
- TensorFlow installation: Pip
- TensorFlow library: 2.4 & 2.4.1

### 2. Code

`
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

def convert_to_lite(model,out_path,enable_selected_tf_ops=False):
    converter = tf.lite.TFLiteConverter.from_keras_model(model)
    out_model = converter.convert()
    with open(out_path, "wb") as fp:
        fp.write(out_model)

def failed_on_android_gpu_for_dynamic_size():
    ipt = layers.Input((256,64,3))
    x = ipt
    x = layers.Reshape((16384,3))(x)
    x = layers.Softmax(axis=1)(x)
    x = layers.Flatten()(x)
    x = layers.Dense(1)(x)
    model = keras.Model(inputs=[ipt],outputs=[x])
    convert_to_lite(model,'failed_on_android_gpu.tflite')

def work_reshape_only():
    ipt = layers.Input((256,64,3))
    x = ipt
    x = layers.Reshape((16384,3))(x)
    # x = layers.Softmax(axis=1)(x)
    x = layers.Flatten()(x)
    x = layers.Dense(1)(x)
    model = keras.Model(inputs=[ipt],outputs=[x])
    convert_to_lite(model,'work_reshape_only.tflite')
[Reshape_before_Softmax_leads_runtime_error.zip](https://github.com/tensorflow/tensorflow/files/5879355/Reshape_before_Softmax_leads_runtime_error.zip)

def work_softmax_only():
    ipt = layers.Input((256,64,3))
    x = ipt
    # x = layers.Reshape((16384,3))(x)
    x = layers.Softmax(axis=1)(x)
    x = layers.Flatten()(x)
    x = layers.Dense(1)(x)
    model = keras.Model(inputs=[ipt],outputs=[x])
    convert_to_lite(model,'work_softmax_only.tflite')

if __name__ == "__main__":
    failed_on_android_gpu_for_dynamic_size()
    work_reshape_only()
    work_softmax_only()
`
### 5. (optional) Any other info / logs
Run the 'failed_on_android_gpu.tflite' on android with gpu delegate cause exception:
Internal error: Failed to apply delegate: Attempting to use a delegate that only supports static-sized tensors with a graph that has dynamic-sized tensors.
The other 2 tflite files which with Reshape/Softmax only works fine, so the dynamic-size tensor is unnecessary.

I also tested nn.softmax or math.reduced_sum, has same problem. Maybe 'reduce_sum_with_dims' is the real 

problem.

**Comment by srjoglekar246** [comment]
In the model with the Reshape & Softmax, there is a sequence on ops (SHAPE -> STRIDED_SLICE -> PACK) that produces the 'shape' input to the Reshape op. Since this input is not a constant, I assume that the Reshape output is marked dynamic in the runtime.

As pointed out by @Robird , this dynamic-ness seems unnecessary.

@karimnosseir , could you take a look and see why the converter might be doing this?

**Comment by karimnosseir** [comment]
Keras by default have dynamic batch size, so the actual input to the model is
[?x256x64x3xf32] and the output is [?x1xf32]

The current converter honor what is passed, so we keep everything dynamic as is and that's why the tensors are dynamic.
If you don't care about the batch size, then you can set batch_size=1 to the input layer.
Example:
```
def failed_on_android_gpu_for_dynamic_size():
  ipt = layers.Input((256, 64, 3), batch_size=1)
  x = ipt
  x = layers.Reshape((16384, 3))(x)
  x = layers.Softmax(axis=1)(x)
  x = layers.Flatten()(x)
  x = layers.Dense(1)(x)
  model = keras.Model(inputs=[ipt], outputs=[x])
  convert_to_lite(model, "failed_on_android_gpu.tflite")
```


Similar question https://github.com/tensorflow/tensorflow/issues/43882#issuecomment-731636562

**Comment by Robird** [comment]
@karimnosseir
